sdk/eidos_sdk/io/events.py

```diff
@@ -0,0 +1,170 @@
+            return AgentStateEvent(**event_dict)
+        else:
+            raise ValueError(f"Unknown event type {event_type}")
+
+
+class StartStreamContextEvent(BaseStreamEvent, ABC):
+    category: Literal[Category.START] = Category.START
+    context_id: str
+
+
+class EndStreamContextEvent(BaseStreamEvent, ABC):
+    category: Literal[Category.START] = Category.END
+    context_id: str
+
+
+class StartLLMEvent(BaseStreamEvent):
+    event_type: Literal["llm"] = "llm"
+    category: Literal[Category.START] = Category.START
+
+
+class ToolCallStartEvent(StartStreamContextEvent):
+    event_type: Literal["tool_call"] = "tool_call_start"
+    tool_call: ToolCall
+
+
+class ToolCallEndEvent(EndStreamContextEvent):
+    event_type: Literal["tool_call"] = "tool_call_end"
+
+
+class StartAgentCallEvent(BaseStreamEvent):
+    category: Literal[Category.START] = Category.START
+    event_type: Literal["agent_call"] = "agent_call"
+    machine: str
+    agent_name: str
+    call_name: str
+    process_id: str
+
+
+class OutputEvent(BaseStreamEvent, ABC):
+    category: Literal[Category.OUTPUT] = Category.OUTPUT
+    content: Any
+
+    @staticmethod
+    def get(content: T, **kwargs):
+        if isinstance(content, str):
+            return StringOutputEvent(content=content, **kwargs)
+        else:
+            return ObjectOutputEvent[T](content=content, **kwargs)
+
+
+class LLMToolCallRequestEvent(BaseStreamEvent):
+    category: Literal[Category.OUTPUT] = Category.OUTPUT
+    event_type: Literal["tool_call"] = "llm_tool_call_request"
+    tool_call: ToolCall
+
+
+class StringOutputEvent(OutputEvent):
+    event_type: Literal["string_output"] = "string"
+    content: str
+
+
+class ObjectOutputEvent(OutputEvent, Generic[T]):
+    event_type: Literal["object_output"] = "object"
+    content: T
+
+
+# note EndStreamEvent does not need to reference the type of event it ends since this is captured by context
+class EndStreamEvent(BaseStreamEvent, ABC):
+    category: Literal[Category.END] = Category.END
+    event_type: StopReason
+
+
+class SuccessEvent(EndStreamEvent):
+    event_type: Literal[StopReason.SUCCESS] = StopReason.SUCCESS
+
+
+class CanceledEvent(EndStreamEvent):
+    event_type: Literal[StopReason.CANCELED] = StopReason.CANCELED
+
+
+class ErrorEvent(EndStreamEvent):
+    event_type: Literal[StopReason.ERROR] = StopReason.ERROR
+    reason: Any
+
+
+class AgentStateEvent(BaseStreamEvent):
+    category: Literal[Category.TRANSFORM] = Category.TRANSFORM
+    event_type: Literal["agent_state"] = "agent_state"
+    state: str
+    available_actions: List[str] = None  # this is filled in by the server, agents should leave the default
+
+
+StreamEvent = StartAgentCallEvent | StartLLMEvent | ToolCallStartEvent | ToolCallEndEvent | LLMToolCallRequestEvent | \
+              StringOutputEvent | ObjectOutputEvent | \
+              SuccessEvent | CanceledEvent | ErrorEvent | \
+              AgentStateEvent
```

Comment:
We have too many events types to reason about. And this is going to blow out as we add more mechanics to the system.

Similarly, I do not think we want to hard code tool calls (and tool calls only) as start context. Imagine a tool call that has two agents executing simultaneously. In this scenario each agent should have its own context.

Here is a proposal to drop it down to 4 event types:

start event (includes optional context if it is new context, also contains an arbitrary data field the tool call can use)
info event (transition and tool calls)
output event
end event